In [ ]:
#importing necessary libraries
import tweepy
import pandas as pd
import networkx as nx
import numpy as np
import collections

In [ ]:
#sign up for twitter developer account to get these credentials

consumer_key=  'XXXXXXXXXXXXXXXXXXXXX'
consumer_secret= 'XXXXXXXXXXXXXXXXXXXXXXXXX'
access_token= 'XXXXXXXXXXXXXXXXXX'
access_token_secret= 'XXXXXXXXXXXXXXXXXXXXXXXXXX'

In [ ]:
#api connection
auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_token_secret)
api = tweepy.API(auth, wait_on_rate_limit=True, wait_on_rate_limit_notify=True, compression=True)

In [ ]:
#to get your own account id -
me = api.get_user(screen_name = 'gaya3cantshutup') #change to your username
me.id

1472558789698019330

In [ ]:
#getting the users that follow me 
user_list = ["1472558789698019330"]
follower_list = []
for user in user_list:
    followers = []
    try:
        for page in tweepy.Cursor(api.followers_ids, user_id=user).pages():
            followers.extend(page)
            print(len(followers))
    except tweepy.TweepError:
        print("error")
        continue
    follower_list.append(followers)

102


### There are 102 followers, but not all are from PES

Manually I have verified that 69 of the 102 are from PES. They will be the first nodes in the network along with me

There can be better ways to verify, but since 102 is a small number, I verified it manually using - https://oauth-playground.glitch.me/?id=findUserById&params=%28%27query%21%28%29%7Ebody%21%28%29%7Epath%21%28%21*%7E**id%21%272244994945%27%29%01*_

In [ ]:
#Or use this code to store id: username is a dictionary and check:

# dict_pes= dict()
# for j in nodes:
#   u = api.get_user(j)
#   dict_pes[j] = u.screen_name

In [ ]:
# filtering out only those in PES
pes_folks= [1424433180753371139,
 1249343092957655042,
 1477144962286243849,
 1196316777279393792,
 1184475769482825728,
 1344738920,
 1133076940842102784,
 1284839845861273602,
 1497605466968723504,
 1497823537679114241,
 958759926549147649,
 1429603854,
 848901692091506688,
 1317498004199256064,
 1277807884861009921,
 1253945298461708291,
 2228020669,
 1483029280699588608,
 1489178779738013698,
 2872401203,
 1468957860,
 1246014522537603072,
 771622961153396736,
 1004642350553968640,
 746327100257271816,
 3274985820,
 1071853071142141952,
 1287594949228150784,
 1192409641273065473,
 473729548,
 1000797725917499393,
 1247213980797427724,
 1278655748768063490,
 883720758723092480,
 902853369332547585,
 1473893833461043203,
 1249295950549401600,
 705763745079791617,
 1382680394349236228,
 1460180698556297221,
 1475691974845886464,
 3034153315,
 1351451066622119937,
 3220242762,
 1357736121535397888,
 1150712699505897473,
 725384293766291456,
 1355748483127369730,
 850556860411269121,
 1303161128873156608,
 1273713079587463173,
 1280037324504526848,
 1248995104817324034,
 802416910813933568,
 1428587673116037121,
 721713204980817927,
 1277637088842784769,
 2469365941,
 1366336069319335943,
 1098839084162789376,
 780439554784366592,
 860040731674775552,
 1289636412921229312,
 1305910070782251009,
 1471504126890557446,
 1168038336541286400,
 3274572859,
 1403983669908303874,
 1436187424506277898]

In [ ]:
#making a df with my id as source and all my followers as target dest
df = pd.DataFrame(columns=['source','target'])
df['target'] = pes_folks
df['source'] = 1472558789698019330 

In [ ]:
#getting my followers' followers
user_list = list(df['target'])
for userID in user_list:
    print(userID)
    followers = []
    follower_list = []

    # fetching the user
    user = api.get_user(userID)
    # fetching the followers_count
    followers_count = user.followers_count

    try:
        for page in tweepy.Cursor(api.followers_ids, user_id=userID).pages():
            followers.extend(page)
            print(len(followers))
            if followers_count >= 250: #Only take first 250 followers to reduce time the script runs
                break
    except tweepy.TweepError:
        print("error")
        continue
    follower_list.append(followers)
    temp = pd.DataFrame(columns=['source', 'target'])
    temp['target'] = follower_list[0]
    temp['source'] = userID
    df = df.append(temp)
    df.to_csv("networkOfFollowers.csv", index=False)

1424433180753371139
44
1249343092957655042
38
1477144962286243849
12
1196316777279393792
23
1184475769482825728
21
1344738920
76
1133076940842102784
25
1284839845861273602
49
1497605466968723504
66
1497823537679114241
50
958759926549147649
81
1429603854
63
848901692091506688


Rate limit reached. Sleeping for: 467


18
1317498004199256064
40
1277807884861009921
65
1253945298461708291
102
2228020669
25
1483029280699588608
107
1489178779738013698
25
2872401203
190
1468957860
101
1246014522537603072
14
771622961153396736
109
1004642350553968640
178
746327100257271816
78
3274985820
102
1071853071142141952
53
1287594949228150784


Rate limit reached. Sleeping for: 896


38
1192409641273065473
34
473729548
77
1000797725917499393
77
1247213980797427724
177
1278655748768063490
128
883720758723092480
52
902853369332547585
139
1473893833461043203
27
1249295950549401600
92
705763745079791617
154
1382680394349236228
48
1460180698556297221
65
1475691974845886464
62
3034153315
188
1351451066622119937


Rate limit reached. Sleeping for: 897


41
3220242762
341
1357736121535397888
57
1150712699505897473
254
725384293766291456
112
1355748483127369730
87
850556860411269121
165
1303161128873156608
58
1273713079587463173
183
1280037324504526848
133
1248995104817324034
139
802416910813933568
179
1428587673116037121
64
721713204980817927
49
1277637088842784769
59
2469365941
143
1366336069319335943


Rate limit reached. Sleeping for: 896


87
1098839084162789376
173
780439554784366592
169
860040731674775552
165
1289636412921229312
108
1305910070782251009
101
1471504126890557446
64
1168038336541286400
66
3274572859
66
1403983669908303874
88
1436187424506277898
96


The followers' followers data has been appended to the pandas df and has been saved to the file networkOfFollowers.csv
After running the code, comment out the above block so that it doesn't run again and use unecessary API get requests

In [ ]:
df_saved=pd.read_csv('/content/networkOfFollowers.csv')
df_check = df_saved[~df_saved.target.isin(pes_folks)]
len(df_check)

4459

In [ ]:
new_people= df_check['target']
counter=collections.Counter(new_people)
unique_new= np.unique(new_people)
len(unique_new)

2472

In [ ]:
counter.most_common(50)

[(1472558789698019330, 65),
 (1242738027249586176, 30),
 (834992149758255104, 27),
 (2838973913, 25),
 (831682449864286208, 24),
 (1300814862130360320, 24),
 (2876732521, 23),
 (745566943080394752, 21),
 (795674214690918400, 21),
 (1277835126236274688, 21),
 (3320026471, 20),
 (833359275233652736, 19),
 (1251386586383527936, 19),
 (1471034166439342096, 19),
 (3221574320, 18),
 (1154274912703356928, 18),
 (1139519706698665984, 17),
 (1379709485325840387, 17),
 (1268952216267902976, 17),
 (1146263752125374464, 17),
 (1282017061573668866, 16),
 (1135751408802680832, 16),
 (4168701552, 16),
 (1104094260263972864, 16),
 (785833074319818752, 16),
 (1148825471338991616, 16),
 (610131860, 15),
 (1305531562243489793, 15),
 (827629337809653761, 14),
 (821685274786856960, 14),
 (1345469474791591936, 14),
 (1457724829465935874, 13),
 (3308507935, 13),
 (1431620243307188224, 13),
 (1139124129188073477, 13),
 (2926993738, 13),
 (1289935140265680897, 13),
 (1250017084890726405, 12),
 (974875971110158

First ID is me, as most people in the pes_folks list also follow me

After this, IDs whose frequency is high, has a high probability of being in PES

We can use any threshold number but I will use 20 as the threshold

Run the code to get followers' followers' followers as shown above and use a threshold of 10 to get final nodes

In [ ]:
#example for one user-
user_list_curr = [1242738027249586176]              
follower_list = []
for user in user_list_curr:
    followers = []
    try:
        for page in tweepy.Cursor(api.followers_ids, user_id=user).pages():
            followers.extend(page)
            print(len(followers))
    except tweepy.TweepError:
        print("error")
        continue
    follower_list.append(followers)

74


In [ ]:
df1 = pd.DataFrame(columns=['source','target'])
df1['target'] = followers #Set the list of followers as the target column
df1['source'] = 1242738027249586176 #Set my user ID as the source 

#need to add df1 to df_saved and continue adding the new followers of all in user_list (friends that are one level away)

In [ ]:
one_level_away=[1472558789698019330,1242738027249586176,834992149758255104,831682449864286208,1300814862130360320,2838973913, \
                2876732521, 745566943080394752,795674214690918400, 1277835126236274688, 3320026471, 1251386586383527936, 3221574320]

In [ ]:
final_pes = pes_folks + one_level_away

In [ ]:
for index, row in df_saved.iterrows():
    if row.target not in final_pes:
        df_saved.drop(index, inplace=True)

In [ ]:
df_saved.to_csv('updated_pes.csv')